File: data-structures/javascript/hash_tables/hash-table.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  addNode(node) {
    if (this.size === 0) {
      let newNode = new Node(node);
      this.head = newNode;
      this.tail = newNode;
      this.size++;
    } else {
      let currentNode = this.head;

      while(currentNode.next !== null) {
        currentNode = currentNode.next;
      }

      currentNode.next = new Node(node);
      this.tail = currentNode.next;
      this.size++;
    }
  }

  contains(data) {
    if (this.head === null) {
      return false;
    }

    if (this.head.data === data || this.tail.data === data) {
      return true;
    } else {
      let currentNode = this.head;

      while(currentNode.next !== null) {
        currentNode = currentNode.next;
        if (currentNode.data === data) {
          return true;
        }
      }
    }

    return false;
  }

  removeNode(data) {
    if (!this.contains(data)) {
      return null;
    }

    let nodeToRemove;

    if (this.head.data === data) {
      nodeToRemove = this.head;
      this.head = this.head.next;
      this.size--;
      return nodeToRemove;
    } else {
      let currentNode = this.head;

      while(currentNode.next.next !== null && currentNode.next.data !== data) {
        currentNode = currentNode.next;
      }

      nodeToRemove = currentNode.next;
      if (this.tail.data === data) {
        this.tail = currentNode;
      }
      currentNode.next = nodeToRemove.next;
      this.size--;
      return nodeToRemove;
    }
  }

  removeHead() {
    if (this.size) {
      let originalHead = this.head;
      this.head = this.head.next;
      this.size--;
      return originalHead;
    }
  }

  removeTail() {
    let originalTail = this.tail;
    let currentNode = this.head;
    if (originalTail === currentNode) {
      this.head.next = null;
      this.head = null;
      this.tail = null;
      this.size--;
      return originalTail;
    }
    while(currentNode.next.data !== this.tail.data) {
      currentNode = currentNode.next;
    }

    currentNode.next = null;
    this.size--;
    this.tail = currentNode;
    return originalTail;
  }

  appendToHead(node) {
    let temp = this.head.next;
    let newNode = new Node(node);
    this.head.next = newNode;
    this.head.next.next = temp;
    this.size++;
  }

  prependToTail(node) {
    let temp = new Node(node);
    let originalTail = this.tail;
    let currentNode = this.head;

    while(currentNode.next.data !== this.tail.data) {
      currentNode = currentNode.next;
    }

    currentNode.next = temp;
    currentNode.next.next = this.tail;
    this.size++;
  }

  cloneLinkedList() {
    let newLinkedList = new LinkedList();
    let currentNode = this.head;

    newLinkedList.addNode(currentNode.data);

    while(currentNode.next !== null) {
      currentNode = currentNode.next;
      newLinkedList.addNode(currentNode.data);
    }

    return newLinkedList;
  }

  mergeLinkedLists(lists) {
    let sortedListNodes = {};
    let sortedLinkedList = new LinkedList();

    lists.forEach(list => {
      while(list.head !== null) {
        let removedFromBack;
        let removedFromFront;

        if (list.tail) {
          removedFromBack = list.removeTail().data;
          sortedListNodes[removedFromBack] === undefined ? sortedListNodes[removedFromBack] = 1 : sortedListNodes[removedFromBack]++;
        }
        if (list.head) {
          removedFromFront = list.removeHead().data;
          sortedListNodes[removedFromFront] === undefined ? sortedListNodes[removedFromFront] = 1 : sortedListNodes[removedFromFront]++;
        }
      }
    });

    for (let node in sortedListNodes) {
      let counter = sortedListNodes[node];
      while(counter > 0) {
        sortedLinkedList.addNode(node);
        counter--;
      }
    }

    return sortedLinkedList;
  }

  listToArray() {
    let result = [this.head.data];

    let current = this.head;

    while(current.next !== null) {
      current = current.next;
      result.push(current.data);
    }

    return result;
  }
}

// need to add resizing of storage
// aka open addressing

// already added closed-addressing
// aka seperational chaining
class HashTable {
  constructor(initialSize) {
    this.storage = [];

    for (let i = 0; i < initialSize; i++) {
      this.storage.push(new LinkedList);
    }

    this.storageLimit = initialSize;
    this.size = 0;

    const constant = 11021992;

    this.x = key => {
      let hash = 0;
      if (typeof key === 'string') {
        for (let i = 0; i < key.length; i++) {
          hash = (hash << 5) + hash + key.charCodeAt(i);
          hash = hash & hash;
          hash = Math.abs(hash << constant);
        }
      } else {
        hash = key << 5;
        hash = hash & hash;
        hash = Math.abs(hash >> constant);
      }

      return hash % this.storageLimit;
    }
  }

  insert(key, val) {
    let bucketIndex = this.hash(key);
    if (this.storage[bucketIndex].head === null) {
      this.storage[bucketIndex].addNode([key, val])
      return true;
    }

    let current = this.storage[bucketIndex].head;
    let complete = false;

    while(current !== null && complete === false) {
      if (current.data[0] === key) {
        current.data[1] = val;
        complete = true;
        return true;
      }
      current = current.next;
    }

    this.storage[bucketIndex].addNode([key, val]);
    return true;
  }


  remove(key) {
    const bucketIndex = this.hash(key);
    if (bucketIndex > this.storageLimit) {
      return false;
    }
    let target;

    const currentBucket = this.storage[bucketIndex];
    let current = currentBucket.head;

    while(current !== null) {
      if (current.data[0] === key) {
        target = current.data;
        current = null;
      } else {
        current = current.next
      }
    }

    if (target) {
      const removed = currentBucket.removeNode(target);
      return removed;
    } else {
      return false;
    }
  }

  get(key) {
    const bucketIndex = this.hash(key);

    if (bucketIndex > this.storageLimit) {
      return false;
    }

    const currentBucket = this.storage[bucketIndex];
    let current = currentBucket.head;
    while(current !== null) {
      if (current.data[0] === key) {
        return current.data;
      }
      current = current.next;
    }

    return false;
  }

  contains(key) {
    return !a.get(key) === false
  }

  hash(key) {
    return this.x(key)
  }
}


let a = new HashTable(10);

console.log(a.hash(78))
a.insert("Doug", "Manager");
console.log(a.get("Doug"));
a.insert("Alex", "Software Engineer");
console.log(a.get("Alex"));
a.insert("Will", "UX/UI Designer");
console.log(a.get("Will"));
a.insert("Steve", "Designer");
console.log(a.get("Steve"));
a.insert("Frank", "Doctor");
console.log(a.get("Frank"));
a.insert("Alex", "Mechanic")
console.log(a.get("Alex"));
console.log(a.get("Q"));
console.log(a.contains("Q"));
console.log(a.contains("Alex"));
console.log(a.remove("Alex"));

File: data-structures/go/hash_tables/hash_table.go
package hash_tables

import (
  "errors"
  "math"
)


type InterfaceLinkedList struct {
  Head *SinglyInterfaceListNode
  Initialized bool
}

type SinglyInterfaceListNode struct {
  Key string
  Val interface{}
  Next *SinglyInterfaceListNode
}

func ConstructInterfaceLinkedList() *InterfaceLinkedList {
  return &InterfaceLinkedList{
    Head: &SinglyInterfaceListNode{},
    Initialized: false,
  }
}

func (ln *InterfaceLinkedList) AppendInt(key string, node interface{}) {
  if ln.Initialized == false {
    ln.Head.Key = key
    ln.Head.Val = node
    ln.Initialized = true
  } else if ln.Head.Key == key {
    ln.Head.Val = node
  } else if ln.Head == nil {
    ln.Head = &SinglyInterfaceListNode{
      Key: key,
      Val: node,
      Next: nil,
    }
  } else {
    curr := ln.Head
    if curr.Key == key {
      curr.Val = node
      return
    }

    for curr.Next != nil {
      curr = curr.Next
    }

    curr.Next = &SinglyInterfaceListNode{
      Key: key,
      Val: node,
      Next: nil,
    }
  }
}

func (ln *InterfaceLinkedList) DeleteInt(key string) *SinglyInterfaceListNode {
  current := ln.Head
	prev := ln.Head
	var removed *SinglyInterfaceListNode

  if current.Key == key && current.Next != nil {
		ln.Head = current.Next
		ln.Head.Next = current.Next.Next
		return current
	} else if current.Key == key && current.Next == nil {
		ln.Head = nil
		return prev
	} else {
		for current.Next != nil {
			prev = current
			current = current.Next
			if current.Key == key {
				prev.Next = current.Next
				removed = current
				return removed
			}
		}

		return removed
	}
}


func (ln *InterfaceLinkedList) ContainsInt(key string) bool {
  if ln == nil {
    return false
  }

  curr := ln.Head
  contained := false

  for curr != nil {
    if curr.Key == key {
      contained = true
      return contained
    } else {
      curr = curr.Next
    }
  }

  return contained
}



type InterfaceHashTable struct {
  Storage []*InterfaceLinkedList
  StorageLimit int
  Size int
}

func constructInterfaceStorage(limit int) []*InterfaceLinkedList {
  storage := make([]*InterfaceLinkedList, 0)
  for i := 0; i < limit; i++ {
    storage = append(storage, ConstructInterfaceLinkedList())
  }

  return storage
}

func ConstructInterfaceHashTable(limit int) *InterfaceHashTable {
  return &InterfaceHashTable{
    Storage: constructInterfaceStorage(limit),
    StorageLimit: limit,
    Size: 0,
  }
}

func (ht *InterfaceHashTable) hashify(key string) int {
  var hash int

  for i := 0; i < len(key); i++ {
		hash = (hash << 5) + hash
		hash = hash + int(rune(string(key[i])[0])-'0')
		hash = hash & hash
		hash = int(math.Abs(float64(uint(hash) << uint(3))))
	}

  return hash % ht.StorageLimit
}

func (ht *InterfaceHashTable) Shrink() *InterfaceHashTable {
  return ht
}

func (ht *InterfaceHashTable) Expand() *InterfaceHashTable {
  return ht
}

func (ht *InterfaceHashTable) Insert(key string, item interface{}) {
  if float64(ht.Size) >= float64(ht.StorageLimit) * 0.65 {
    ht = ht.Expand()
  }

  idx := ht.hashify(key)
  ht.Storage[idx].AppendInt(key, item)
  ht.Size++
}

func (ht *InterfaceHashTable) Remove(key string) (*SinglyInterfaceListNode, error) {
  if float64(ht.Size) > float64(ht.StorageLimit) * 0.25 && float64(ht.Size) <= float64(ht.StorageLimit) * 0.3 {
    ht = ht.Shrink()
  }

  idx := ht.hashify(key)
  if idx >= ht.StorageLimit {
    return nil, errors.New("Key does not exist")
  }

  removed := ht.Storage[idx].DeleteInt(key)
  ht.Size--
  return removed, nil
}

func (ht *InterfaceHashTable) Contains(key string) bool {
  idx := ht.hashify(key)
  if (idx >= ht.StorageLimit) == true {
    return false
  }

  return ht.Storage[idx].ContainsInt(key)
}
